keps/sig-cli/3805-ssa-default/README.md

```diff
@@ -0,0 +1,561 @@
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+#### Alpha
+
+Alpha is the current level of the feature since server-side apply is
+currently enabled by default in Kubernetes, but enabled on-demand by
+kubectl users.
+
+The feature already has a fair amount of usage since tools (sometimes
+outside of kubectl) have used it both as "clients" and in controllers.
+
+#### Beta
+
+<!--
+To be re-evaluated later:
+Server-Side Apply has a very limited set of bugs or feature requests as
+this point and is definitely mature. Enabling client-side will allow
+increased usage and reduce burden cost for kubectl to maintain both
+mechanisms. -->
+
+#### GA
+
+<!--
+To be re-evaluated later:
+Kubectl doesn't have real-time metrics for usage. The decision to move
+to server-side entirely by default (if ever enabled) will be driven by
+bug reports and complaints from customers. Also by the ability to
+migrate existing client-side usage to server-side. -->
+
+#### Deprecation
+
+We are not intending to deprecate the flag, but we might remove the
+`--server-side` flag in the long term.
+
+Same thing applies for `--save-config` and other client-side related
+flags in kubectl which we might remove.
+
+### Upgrade / Downgrade Strategy
+
+While upgrade / downgrade doesn't really apply to a kubectl feature, we
+currently have a upgrade (and somewhat downgrade) feature in kubectl to
+go from client-side to server-side apply. The upgrade and downgrade
+works well in the nominal cases but fail with special cases. Enabling
+server-side by default also intends to address that problem.
+
+### Version Skew Strategy
+
+N/A.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Other
+  - Describe the mechanism:
+
+Feature is enabled on-demand, and can stop using it at any time.
```

Comment:
No, that's not sufficient. It needs to be guarded by an env variable as we do with CLI features. In other words, the user must *both* set the ENV var AND use the new option value.

This is how the user positively acknowledges that they know they are using an alpha feature. Otherwise, they think that flag option is stable.

Reply:
@johnbelamaric This would at least partially destroy the point of the the KEP and risk pushing the already ridiculously slow timeline back another 3 releases ~1y? The flag switches between two already stable features. Please consider an exception.

Alternatively they could make it JUST be an env var, no flag change needed, but that is a weird user experience.

Reply:
That's not really all it does, because it automatically determines which one to use, based on a guess as to which one was used in the past. We don't know if that guessing algorithm has unintended consequences.

Here's a possible compromise. The user has to use this value explicitly. Which means they need to find it in the usage output or docs. Can we just add in those places a note that it is alpha and to use it they must both use that value and set the env var?

Yes, it makes it a little more awkward to use, but not substantially so from the point of view of a user that it trying it for the first time. The advantage is that it also means that copy-paste of `kubectl` commands all over the internet that have `auto` in them will have to include the env var and it becomes much clearer that this is not a fully baked thing yet.